builtin/credential/approle/path_tidy_user_id.go

```diff
@@ -34,98 +34,112 @@ func (b *backend) tidySecretID(ctx context.Context, s logical.Storage) error {
 		return fmt.Errorf("SecretID tidy operation already running")
 	}
 
-	roleNameHMACs, err := s.List(ctx, "secret_id/")
-	if err != nil {
-		return err
-	}
+	var result error
 
-	// List all the accessors and add them all to a map
-	accessorHashes, err := s.List(ctx, "accessor/")
-	if err != nil {
-		return err
-	}
-	accessorMap := make(map[string]bool, len(accessorHashes))
-	for _, accessorHash := range accessorHashes {
-		accessorMap[accessorHash] = true
-	}
+	tidyFunc := func(secretIDPrefixToUse, accessorIDPrefixToUse string) error {
+		roleNameHMACs, err := s.List(ctx, secretIDPrefixToUse)
+		if err != nil {
+			return err
+		}
 
-	var result error
-	for _, roleNameHMAC := range roleNameHMACs {
-		// roleNameHMAC will already have a '/' suffix. Don't append another one.
-		secretIDHMACs, err := s.List(ctx, fmt.Sprintf("secret_id/%s", roleNameHMAC))
+		// List all the accessors and add them all to a map
+		accessorHashes, err := s.List(ctx, accessorIDPrefixToUse)
 		if err != nil {
 			return err
 		}
-		for _, secretIDHMAC := range secretIDHMACs {
-			// In order to avoid lock swroleing in case there is need to delete,
-			// grab the write lock.
-			lock := b.secretIDLock(secretIDHMAC)
-			lock.Lock()
+		accessorMap := make(map[string]bool, len(accessorHashes))
+		for _, accessorHash := range accessorHashes {
+			accessorMap[accessorHash] = true
+		}
+
+		for _, roleNameHMAC := range roleNameHMACs {
 			// roleNameHMAC will already have a '/' suffix. Don't append another one.
-			entryIndex := fmt.Sprintf("secret_id/%s%s", roleNameHMAC, secretIDHMAC)
-			secretIDEntry, err := s.Get(ctx, entryIndex)
+			secretIDHMACs, err := s.List(ctx, fmt.Sprintf("%s%s", secretIDPrefixToUse, roleNameHMAC))
 			if err != nil {
-				lock.Unlock()
-				return errwrap.Wrapf(fmt.Sprintf("error fetching SecretID %q: {{err}}", secretIDHMAC), err)
-			}
-
-			if secretIDEntry == nil {
-				result = multierror.Append(result, fmt.Errorf("entry for SecretID %q is nil", secretIDHMAC))
-				lock.Unlock()
-				continue
+				return err
 			}
+			for _, secretIDHMAC := range secretIDHMACs {
+				// In order to avoid lock swroleing in case there is need to delete,
+				// grab the write lock.
+				lock := b.secretIDLock(secretIDHMAC)
+				lock.Lock()
+				// roleNameHMAC will already have a '/' suffix. Don't append another one.
+				entryIndex := fmt.Sprintf("%s%s%s", secretIDPrefixToUse, roleNameHMAC, secretIDHMAC)
+				secretIDEntry, err := s.Get(ctx, entryIndex)
+				if err != nil {
+					lock.Unlock()
```

Comment:
An alternative way to deal with unlocking would be to put the interior of the loop into its own function, and then to defer unlocking within that method.